swarm.go

```diff
@@ -168,33 +165,46 @@ func (s *Swarm) teardown() error {
 	return nil
 }
 
-// AddAddrFilter adds a multiaddr filter to the set of filters the swarm will use to determine which
-// addresses not to dial to.
-func (s *Swarm) AddAddrFilter(f string) error {
-	m, err := mafilter.NewMask(f)
-	if err != nil {
-		return err
-	}
-
-	s.Filters.AddDialFilter(m)
-	return nil
-}
-
 // Process returns the Process of the swarm
 func (s *Swarm) Process() goprocess.Process {
 	return s.proc
 }
 
 func (s *Swarm) addConn(tc transport.CapableConn, dir network.Direction) (*Conn, error) {
-	// The underlying transport (or the dialer) *should* filter it's own
-	// connections but we should double check anyways.
-	raddr := tc.RemoteMultiaddr()
-	if s.Filters.AddrBlocked(raddr) {
-		tc.Close()
-		return nil, ErrAddrFiltered
+	p := tc.RemotePeer()
+	// last fallback check for gating connections
+	var rejectConnection bool
+	if s.ConnGater != nil {
+		switch dir {
+		case network.DirInbound:
+			if !s.ConnGater.InterceptAccept(tc) {
+				rejectConnection = true
+			}
+		case network.DirOutbound:
+			if !s.ConnGater.InterceptPeerAddrDial(p, tc.RemoteMultiaddr()) || !s.ConnGater.InterceptPeerDial(p) {
```

Comment:
Why do we intercept this after the connection has been established? Can't we intercept it before we make the connection attempt (since we already know the target peer ID and the multiaddr we're about to dial)?

Reply:
This has been removed.